subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -35,13 +35,33 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+/**
+ * Updater for hierarchical file watchers.
+ *
+ * This is the lifecycle for the watched project root directories:
+ * - During a build, there will be various calls to {@link #updateProjectRootDirectories(Collection)},
+ *   each call augmenting the collection. The watchers will be updated accordingly.
+ * - We try not to stop watching project root directories during a build.
+ * - At the end of the build
+ *   - stop watching the project root directories with nothing to watch inside
+ *   - remember the current watched project root directories as old root directories for the next build
+ *   - remove all non-watched project root directories from the old root directories.
+ * - When updating the watches, we watch project root directories or old project root directories instead of
+ *   directories inside them.
+ *
+ * The goal of the above logic is to
+ * - Keep the updates to the watched directories during a build to a minimum.
+ * - Release the watch on a project directory if it is deleted while the daemon is idle.
```

Comment:
I'd actually move this section about the goals before the explanation of the lifecycle.